### DISCUSSION TITLE
Is there an easy way to create an update pattern from the table struct?

### DISCUSSION THREAD

**Discussion by cubicYYY** [diesel-rs/diesel#3866] [Q&A] [answered]
I often encountered with a pattern like this:
```rust
#[derive(
    ToSchema, Serialize, Deserialize, Selectable, Queryable, Insertable, Identifiable, Clone, Debug,
)]
#[diesel(table_name = crate::schema::device)]
#[diesel(check_for_backend(Mysql))]
pub struct Device {
    pub id: u64,
    pub uid: u64,
    pub name: String,
    pub desc: Option<String>,
    pub dtype: u32,
    pub latitude: Option<f64>,
    pub longitude: Option<f64>,
    #[serde(with = "ts_milliseconds")]
    pub since: NaiveDateTime,
    #[serde(with = "ts_milliseconds")]
    pub last_update: NaiveDateTime,
    pub activated: bool,
}

#[derive(ToSchema, AsChangeset, Clone, Debug)]
#[diesel(table_name = crate::schema::device)]
#[diesel(check_for_backend(Mysql))]
pub struct UpdateDevice<'a> {
    pub uid: Option<u64>,
    pub name: Option<&'a str>,
    pub desc: Option<Option<&'a str>>,
    pub latitude: Option<Option<f64>>,
    pub longitude: Option<Option<f64>>,
    pub last_update: Option<&'a NaiveDateTime>,
    pub activated: Option<bool>,
}
```
You can see I have to wrap tons of `Option<>` on the fields in the struct for updating, since all the fields (except for the primary key) can be not updated (i.e. keep unchanged) in an update query.
How to simplify this?

**Comment by weiznich** [comment, marked as answer]
It's a usual pattern in strongly typed languages to have distinct types for different usecases. In this case your usecases are loading and updating data, which are different operations that expect different data structures. You can see that by the fact that your structs have different field types, also if you look at the inner types of the option. 
It might look like a lot of duplicated code at the first look, but this pattern enables you to evolve the different API's (load and update in this case) independently in the future.
